Fix #2348, add handling for non recording span in jaeger propagator

bluesliverx opened the pull request:
Fixes #2348

# Description

This is a simple fix for #2348, where the jaeger exporter fails to handle a NonRecordingSpan correctly. It just basically skips the parent ID retrieval and moves along as if nothing else happened. This only occurs when a NonRecordingSpan has a valid context, which apparently we have trigged periodically in our app.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

I created a new unit test inspired by an existing test, but done with a NonRecordingSpan to ensure that this doesn't cause the process to fail.

# Does This PR Require a Contrib Repo Change?

Answer the following question based on these examples of changes that would require a Contrib Repo Change:
- [The OTel specification](https://github.com/open-telemetry/opentelemetry-specification) has changed which prompted this PR to update the method interfaces of `opentelemetry-api/` or `opentelemetry-sdk/`
- The method interfaces of `test/util` have changed
- Scripts in `scripts/` that were copied over to the Contrib repo have changed
- Configuration files that were copied over to the Contrib repo have changed (when consistency between repositories is applicable) such as in
    - `pyproject.toml`
    - `isort.cfg`
    - `.flake8`
- When a new `.github/CODEOWNER` is added
- Major changes to project information, such as in:
    - `README.md`
    - `CONTRIBUTING.md`

- [ ] Yes. - Link to PR: 
- [x] No.

# Checklist:

- [x] Followed the style guidelines of this project
- [n/a] Changelogs have been updated
- [x] Unit tests have been added
- [n/a] Documentation has been updated

bluesliverx commented:
I have requested to be added to the approved contributor list for Adobe for the CLA, will hopefully hear back in the next day or so on that.

ocelotl commented on propagator/opentelemetry-propagator-jaeger/src/opentelemetry/propagators/jaeger/__init__.py:
```diff
@@ -79,9 +79,11 @@ def inject(
         span = trace.get_current_span(context=context)
         span_context = span.get_span_context()
         if span_context == trace.INVALID_SPAN_CONTEXT:
+            print(f'invalid span context')
```
We don't use `print` to inform the user of a potential problem, but use a logger with a warning message.

bluesliverx replied:
Sigh, this is why you shouldn't work late 🤦 . This was just for debugging, I've removed it now.

ocelotl commented on propagator/opentelemetry-propagator-jaeger/src/opentelemetry/propagators/jaeger/__init__.py:
```diff
@@ -79,9 +79,11 @@ def inject(
 
-        span_parent_id = span.parent.span_id if span.parent else 0
+        # NonRecordingSpans do not have a parent, skip injecting into them
+        span_parent_id = span.parent.span_id if getattr(span, 'parent', None) else 0
```
Better to use [`hasattr`](https://docs.python.org/3/library/functions.html#hasattr) to check if an object has an attribute or not. Also, is the intention to `skip injecting` into the span if it does not have a parent? This is just setting `span_parent_id` to `0`.

bluesliverx replied:
I was just trying to simplify the check. Originally I was skipping the injection but ended up just keeping the original behavior with a parent of `0`. I updated the comment.

ocelotl commented on propagator/opentelemetry-propagator-jaeger/tests/test_jaeger_propagator.py:
```diff
@@ -230,3 +230,15 @@ def test_extract_invalid_uber_trace_id_header_to_implicit_ctx(self):
                 self.assertDictEqual(Context(), ctx)
+    
+    def test_non_recording_span(self):
+        """Make sure propagator does not crash when working with NonRecordingSpan"""
```
This should describe what is being tested, and what is being tested is what is being asserted. In this case what is being asserted is the equality of `FORMAT.fields` and `inject_fields`. I understand that this test will fail if the propagator "crashes" somehow, but that is not how "not crashing" should be tested. If what you want to test is that nothing "crashes" (with that I mean that no exception is raised during the execution of the test) then the test should have this form:

```python
    def test_something(self):
        try:
            # The piece of code that needs to be tested for not crashing goes here.
            ...
        except Exception as error:
            self.fail(f"The code being tested failed with {error}")
```

bluesliverx replied:
Updated this test.

bluesliverx commented:
Thanks for the review @ocelotl, I've updated the PR based on your comments.

bluesliverx commented on propagator/opentelemetry-propagator-jaeger/src/opentelemetry/propagators/jaeger/__init__.py:
```diff
@@ -81,7 +81,8 @@ def inject(
 
-        span_parent_id = span.parent.span_id if span.parent else 0
+        # NonRecordingSpans do not have a parent, set parent to 0 in this case
+        span_parent_id = span.parent.span_id if hasattr(span, 'parent') and span.parent else 0
```
Pushed a change for this, let me know if that works.

bluesliverx commented:
I realized EasyCLA was telling me that I had the wrong user (my work email) instead of my private email on the commit. I went and and just combined them all into a single commit and changed the author on them. Things should be good to go now. Not sure if it needs re-approvals since I rewrote history or not. @ocelotl @lzchen

bluesliverx commented:
Ran into pylint errors, I believe I have those all fixed now if you take another look. Sorry for the extra trouble.
